### PR TITLE
15 feature 42 create use cases and persistence

### LINKED ISSUE #15: [Feature 4.2] Create use cases and persistence
Create the delete, edit and send message use cases and persistence

### PR DESCRIPTION
**Pull request by EricYH2134**
Added EditMessage, SendMessage, and DeleteMessage use cases and ChatRepository and MessageRepository files where it saves to JSON persistence, along with interfaces and request models necessary to facilitate it in accordance to clean architecture. Also added test cases for each use case

### CODE REVIEW

**Review by rryt000 — APPROVED** [review]
Overall, the Chat repository and use cases are looking great. As far as my knowledge of clean architecture goes, all the elements such as the Presenter, Gateway, Output and Input Boundaries, and response models are present. One thing to pay attention to is perhaps complete the overriden methods in the ChatPresenter class, and consider catching all the exceptions thrown (this may entail creating custom exceptions). We should also combine our repositories together, since they overlap in quite a few aspects. Other than that 👍

**Review by nik-uoft — APPROVED** [review]
A few minor comments, otherwise LGTM

File: src/main/java/chat/ChatRepository.java [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+    private int numChats;
+
+    public ChatRepository(String filePath) throws IOException {
+        this.filePath = filePath;
+        File JSONFile = new File(filePath);
+        this.numChats = 0;
+        if (JSONFile.length() > 0) {
+            FileReader reader = new FileReader(JSONFile);
+            Gson gson = new GsonBuilder().create();
+            chats = gson.fromJson(reader, new TypeToken<HashMap<Integer, ChatRepoRequestModel>>() {}.getType());
```
Comment (nik-uoft) [inline src/main/java/chat/ChatRepository.java:30]: nit: let's use `this` to disambiguate variable references.

File: src/main/java/chat/ChatRequestModel.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+    private String content;
+
+    private int author;
+
+    private int receiver;
+
+    private Date sendTime;
+
+    public ChatRequestModel(int chatId, String content, int author, int receiver, Date sendTime) {
```
Comment (nik-uoft) [inline src/main/java/chat/ChatRequestModel.java:17]: nit: for `ChatRequestModel` maybe worth to use a single pair of getter and setter?

File: src/main/java/chat/ChatResponseModel.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+    private String content;
+
+    private int author;
+
+    private int receiver;
+
+    private Date sendTime;
+
+    public ChatResponseModel(String content, int author, int receiver, Date sendTime) {
```
Comment (nik-uoft) [inline src/main/java/chat/ChatResponseModel.java:15]: nit: for `ChatResponseModel` maybe worth to use a single pair of getter and setter?

**Review by PotatoRations — APPROVED** [review]
Looks good! Elements of clean architecture are present and done (to my knowledge) correctly. There are entities, respository, use cases, and interfaces separating each layer when needed.

Some considerations and code smell imrpovements: 
1. You'll likely need to catch exceptions from your repository in your use case instead of throwing it in your use case, as there is nowhere further up to catch them. Catching an exception can trigger the failview method. 
2. We will need to establish a common naming convention for our classes, this can be discussed later when we integrate our parts together
3. You have a lot of getters and setters, some that are not used. Interface segregation principle suggests that we should limit the number of public methods. This can be refactored later. 

I will approve as the issues are not significant and can be fixed later when combining and refactoring

**Pull request merged**